yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts

```diff
@@ -0,0 +1,210 @@
+   * @returns The last used index.
+   */
+  async getLastUsedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    const lastFinalizedIndex = await this.#lastFinalizedIndexes.getAsync(secret.toString());
+    const pendingTxScopedIndexes = (await this.#pendingIndexes.getAsync(secret.toString())) ?? [];
+    const pendingIndexes = pendingTxScopedIndexes.map(entry => entry.index);
+
+    if (pendingTxScopedIndexes.length === 0) {
+      return lastFinalizedIndex;
+    }
+
+    // As the last used index we return the highest one from the pending indexes. Note that this might not technically
+    // be the last used index in case we sent logs from multiple PXEs in parallel but that's just a detail.
+    return Math.max(...pendingIndexes);
+  }
+
+  /**
+   * A tx that contained private logs with tags corresponding to some of the indexes returned from this data provider
+   * has been dropped so we delete the corresponding pending indexes. This results in the indexes being reused which
+   * results in the txs that reused the indexes to be linked with the dropped tx. We don't worry about that for now.
+   * @param txHash - The hash of the tx to drop the pending indexes for.
+   */
+  async dropPendingIndexes(txHash: TxHash) {
+    const txHashStr = txHash.toString();
+    const allSecrets = await toArray(this.#pendingIndexes.keysAsync());
+
+    for (const secret of allSecrets) {
+      const pendingData = await this.#pendingIndexes.getAsync(secret);
+      if (pendingData) {
+        const filtered = pendingData.filter(item => item.txHash.toString() !== txHashStr);
+        if (filtered.length === 0) {
+          await this.#pendingIndexes.delete(secret);
+        } else if (filtered.length !== pendingData.length) {
+          // Some items were filtered out, so update the pending data
+          await this.#pendingIndexes.set(secret, filtered);
+        }
+        // else: No items were filtered out (txHash not found for this secret) --> no-op
+      }
+    }
+  }
+
+  /**
+   * A tx that contained private logs with tags corresponding to some of the indexes returned from this data provider
+   * has been finalized so we update the corresponding pending indexes to be finalized (we move them from pending map
+   * to finalized map).
+   * @param txHash - The hash of the tx to update the status of.
+   */
+  async updateStatusToFinalized(txHash: TxHash) {
+    const txHashStr = txHash.toString();
+    const allSecrets = await toArray(this.#pendingIndexes.keysAsync());
+
+    for (const secret of allSecrets) {
+      const pendingData = await this.#pendingIndexes.getAsync(secret);
+      if (pendingData) {
+        const matchingIndexes = pendingData
+          .filter(item => item.txHash.toString() === txHashStr)
+          .map(item => item.index);
+        if (matchingIndexes.length === 1) {
+          let currentFinalized = await this.#lastFinalizedIndexes.getAsync(secret);
```

Comment:
I had to read this repeatedly as i was getting confused by it. This whole section (`matching.len == 1`) is complicated enough to go into its own function. Some comments:
 - `currentFinalized` is a confusing name for something we also call `lastFinalized`.
 - the pruning is not well explained. In particular, it doesn't mention that even though all pending entries should be larger than the last finalized when they were created (we reject smaller ones), we don't control the order in which entries get finalized, and so discard these
 - given the above prune, it should be impossible for us to finalize an index that is smaller than the last one we saw (because we've pruned those entries)